lib/libvarnish/vfc.c

```diff
@@ -41,95 +42,105 @@
 
 #include "config.h"
 
+#include <pthread.h>
+#include <stdint.h>
 #include <stdlib.h>
 #include <string.h>
 
-#include "cache/cache.h"
-
+#include "vdef.h"
+#include "vas.h"
 #include "vfil.h"
+#include "vqueue.h"
+#include "miniobj.h"
 
-#include "vcc_if.h"
+#include "vrt.h" /* XXX: maybe VFC belongs in cache_varnishd.h */
+#include "vfc.h"
 
-struct frfile {
-	unsigned			magic;
-#define CACHED_FILE_MAGIC 0xa8e9d87a
-	char				*file_name;
-	char				*contents;
-	int				refcount;
-	VTAILQ_ENTRY(frfile)		list;
-};
+static VTAILQ_HEAD(, vfc)	vfc_head = VTAILQ_HEAD_INITIALIZER(vfc_head);
+static pthread_mutex_t		vfc_mtx = PTHREAD_MUTEX_INITIALIZER;
 
-static VTAILQ_HEAD(, frfile)	frlist = VTAILQ_HEAD_INITIALIZER(frlist);
-static pthread_mutex_t		frmtx = PTHREAD_MUTEX_INITIALIZER;
+static void
+vfc_unref_locked(struct vfc **vfcp)
+{
+	struct vfc *vfc;
+
+	AN(vfcp);
+	vfc = *vfcp;
+	if (vfc == NULL)
+		return;
+	CHECK_OBJ(vfc, CACHED_FILE_MAGIC);
+	assert(vfc->refcount > 0);
+	if (--vfc->refcount > 0)
+		*vfcp = NULL;
+	else
+		VTAILQ_REMOVE(&vfc_head, vfc, list);
+}
 
 static void
-free_frfile(void *ptr)
+vfc_destroy(struct vfc **vfcp)
 {
-	struct frfile *frf;
+	struct vfc *vfc;
+
+	TAKE_OBJ_NOTNULL(vfc, vfcp, CACHED_FILE_MAGIC);
+	free(TRUST_ME(vfc->contents->blob));
+	free(vfc->file_name);
+	FREE_OBJ(vfc);
+}
 
-	CAST_OBJ_NOTNULL(frf, ptr, CACHED_FILE_MAGIC);
+void
+VFC_destroy(struct vfc **vfcp)
+{
 
-	AZ(pthread_mutex_lock(&frmtx));
-	if (--frf->refcount > 0)
-		frf = NULL;
-	else
-		VTAILQ_REMOVE(&frlist, frf, list);
-	AZ(pthread_mutex_unlock(&frmtx));
-	if (frf != NULL) {
-		free(frf->contents);
-		free(frf->file_name);
-		FREE_OBJ(frf);
-	}
+	AN(vfcp);
+	AN(*vfcp);
+	AZ(pthread_mutex_lock(&vfc_mtx));
+	vfc_unref_locked(vfcp);
+	AZ(pthread_mutex_unlock(&vfc_mtx));
+	if (*vfcp != NULL)
+		vfc_destroy(vfcp);
 }
 
-VCL_STRING v_matchproto_(td_std_fileread)
-vmod_fileread(VRT_CTX, struct vmod_priv *priv,
-    VCL_STRING file_name)
+struct vfc *
+VFC_find(struct vfc *old, const char *file_name)
```

Comment:
I am aware that the purpose of this is to share code between the `std` and `blob` vmods.
Yet I really do not like this to end up as an official API because it just caches the contents forever and provides no means no to. Also I provides no checks for updates etc. etc.
So _if_ we really make this move, I think we should at least mark the interface as deprecated and transient.

Before I go about enumerating all the deficits, I would rather want to point interested readers to [the UPLEX vmod_file](https://code.uplex.de/uplex-varnish/libvmod-file) where we attempt to address issues around caching, updates and file deletion.